### PR TITLE
Mixed precision `amip_interp_mod`

### PR DESCRIPTION
**Pull request by J-Lentz**
**Description**
Mixed precision `amip_interp_mod`

**How Has This Been Tested?**
Builds with Intel on the AMD box.

**Checklist:**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules
- [x] New check tests, if applicable, are included
- [x] `make distcheck` passes

### CODE REVIEW

File: amip_interp/include/amip_interp_r4.fh [hunk 1]
```diff
@@ -0,0 +1,71 @@
+#undef AMIP_INTERP_KIND_
```
Comment (mlee03) [inline amip_interp/include/amip_interp_r4.fh:1]: FMS_AMIP_INTERP_KIND_? :)

File: amip_interp/amip_interp.F90 [hunk 2]
```diff
@@ -265,31 +274,48 @@ module amip_interp_mod
...
 
 !----- public data type ------
 
-!> @brief Contains information needed by the interpolation module (exchange_mod) and buffers data.
+!> @brief Contains information needed by the interpolation module (exchange_mod) and buffers
+!! data. Real precision is determined by the rkind parameter, which defaults to r8_kind.
 !> @ingroup amip_interp_mod
-type amip_interp_type
+type amip_interp_type(rkind)
+   integer, kind            :: rkind = r8_kind ! Real kind for data1 and data2
```
Comment (mlee03) [inline amip_interp/amip_interp.F90:302]: I think we want to avoid using parameterized derived types for now

File: amip_interp/include/amip_interp.inc [hunk 3]
```diff
@@ -16,382 +16,44 @@
...
    type (time_type),         intent(in)    :: Time !< Time to interpolate
-   type (amip_interp_type),  intent(inout) :: Interp !< Holds data for interpolation
-   real,                     intent(out)   ::  sst(:,:) !< Sea surface temperature data
+   type (AMIP_INTERP_TYPE_),  intent(inout) :: Interp !< Holds data for interpolation
+   real(AMIP_INTERP_KIND_),                     intent(out)   ::  sst(:,:) !< Sea surface temperature data
    character(len=*), optional, intent(out) :: err_msg !< Holds error message string if present
 
-   real, dimension(mobs,nobs) :: sice
+   real(AMIP_INTERP_KIND_), dimension(mobs,nobs) :: sice
+   real(AMIP_INTERP_KIND_), allocatable:: temp1(:,:), temp2(:,:)
```
Comment (mlee03) [inline amip_interp/include/amip_interp.inc:29]: is it safe to move temp1 and temp2 to here?

File: amip_interp/amip_interp.F90 [hunk 4]
```diff
@@ -146,18 +146,17 @@ module amip_interp_mod
...
 !-----------------------------------------------------------------------
 !----------------- Public Data -----------------------------------
 integer :: i_sst = 1200
 integer :: j_sst = 600
-real, parameter:: big_number = 1.E30
+real(r8_kind), parameter:: big_number = 1.E30
 logical :: forecast_mode = .false.
-real, allocatable, dimension(:,:) ::  sst_ncep, sst_anom
-
-public i_sst, j_sst, sst_ncep, sst_anom, forecast_mode, use_ncep_sst
```
Comment (mlee03) [inline amip_interp/amip_interp.F90:160]: not sure about changing the public variable names or making some of these private  in mixedmode.  Perhaps these changes could be proposed to main instead.

**Review by mlee03 — CHANGES_REQUESTED** [review]

File: amip_interp/amip_interp.F90 (hunk 2 above)
Comment (J-Lentz) [reply amip_interp/amip_interp.F90]: I've replaced `amip_interp_type` with `amip_interp_type_r4` and `amip_interp_type_r8` in 323ef7b7da84b0ae73e535c9adf2a20e86a332ee. I don't think there's a way to export the name `amip_interp_type` as an alias for `amip_interp_type_r8`, so minor changes to the model code will be required. Fortunately, I believe `ice_amip` and `ocean_amip` are the only model codes which use this module.

File: amip_interp/include/amip_interp_r4.fh (hunk 1 above)
Comment (J-Lentz) [reply amip_interp/include/amip_interp_r4.fh]: Renamed in a15e58ea3fb062be330ce3e29219d7f1b6f959fa.

File: amip_interp/include/amip_interp.inc (hunk 3 above)
Comment (J-Lentz) [reply amip_interp/include/amip_interp.inc]: I think so, but just to err on the side of caution, I've added the `save` attribute in 3e0db621c989b957ad4506fa4abcfae1e8869b2e.

File: amip_interp/amip_interp.F90 (hunk 4 above)
Comment (J-Lentz) [reply amip_interp/amip_interp.F90:160]: One alternative could be to create `sst_ncep` and `sst_anom` pointers as aliases for the r8 versions, but this seems unnecessary if no model code is actually using them.

Comment (mlee03) [reply amip_interp/amip_interp.F90:160]: They could be used by non-GFDL models, @bensonr @laurenchilutti?

**Comment by bensonr** [comment]
amip_interp_mod is used in external_sst.F90
<https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/blob/main/tools/external_sst.F90>
in all branches and GFDL/atmosphere.F90
<https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/blob/main/driver/GFDL/atmosphere.F90>
for GFDL climate models.

File: amip_interp/amip_interp.F90 (hunk 4 above)
Comment (bensonr) [reply amip_interp/amip_interp.F90:160]: @mlee03 @laurenchilutti - those public entities from amip_interp_mod are used in [external_sst.F90](https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/blob/main/tools/external_sst.F90) in all branches and [GFDL/atmosphere.F90](https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/blob/main/driver/GFDL/atmosphere.F90) for GFDL climate models.

File: amip_interp/include/amip_interp.inc [hunk 5]
```diff
@@ -418,7 +80,7 @@ if ( .not.use_daily ) then
    if ( .not. allocated(temp2) ) allocate (temp2(mobs,nobs))
 
    if (use_zonal) then
-      call zonal_sst (Amip_Time, sice, temp1)
+      call ZONAL_SST_ (Amip_Time, sice, temp1)
```
Comment (mlee03) [inline amip_interp/include/amip_interp.inc:83]: Nice, this seems to be better than declaring bunch of generic interfaces for private procedures

File: amip_interp/amip_interp.F90 (hunk 4 above)
Comment (mlee03) [reply amip_interp/amip_interp.F90:160]: @bensonr thank you, @J-Lentz would you be able to change these back to their original names? :)

**Review by mcallic2 — APPROVED** [review]

File: amip_interp/amip_interp.F90 (hunk 4 above)
Comment (J-Lentz) [reply amip_interp/amip_interp.F90:160]: Sorry about that, I'd gotten the impression that this module is only used by `ice_amip` and `ocean_amip`. d8d70837fea3f26e79da605ff45ca0abcfd34515 restores the public data and adds public pointers to the r8 versions of `sst_ncep` and `sst_anom`.

File: amip_interp/amip_interp.F90 [hunk 6]
```diff
@@ -265,31 +280,59 @@ module amip_interp_mod
...
+end interface amip_interp_del
 
 !----- public data type ------
 
-!> @brief Contains information needed by the interpolation module (exchange_mod) and buffers data.
+!> @brief Contains information needed by the interpolation module (exchange_mod) and buffers
+!! data (r4_kind flavor).
 !> @ingroup amip_interp_mod
-type amip_interp_type
+type amip_interp_type_r4
```
Comment (mlee03) [inline amip_interp/amip_interp.F90:307]: Sorry didn't realize amp_interp_type was public.  Is there anyway to mixed precision to amip_interp without changing the name of amip_interp_type?

File: amip_interp/include/amip_interp.inc [hunk 7]
```diff
@@ -452,25 +114,25 @@ if ( .not.use_daily ) then
...
       endif
 
 !-----------------------------------------------------------------------
 
       if (Date2 /= Interp % Date2) then
-          call read_record ('sst', Date2, Udate2, temp2)
+          call READ_RECORD_ ('sst', Date2, Udate2, temp2)
           if ( use_ncep_sst .and. (.not. no_anom_sst) ) then
-               temp2(:,:) = temp2(:,:) + sst_anom(:,:)
+               temp2(:,:) = temp2(:,:) + SST_ANOM_ (:,:)
```
Comment (mlee03) [inline amip_interp/include/amip_interp.inc:132]: For the case where sst_anom_r4 is used... that means sst_anom_r8 is not used and the pointer sst_anom is pointing to sst_anom_r8, how does the user retrieve the sst_anom_r4 data?

File: amip_interp/amip_interp.F90 (hunk 6 above)
Comment (J-Lentz) [reply amip_interp/amip_interp.F90:307]: This could be done by making `data1` and `data2` both `class(*)` and wrapping all code that uses them in `select type` blocks.

File: CMakeLists.txt [hunk 8]
```diff
@@ -323,7 +323,8 @@ foreach(kind ${kinds})
                                                  tracer_manager/include
                                                  interpolator/include
                                                  coupler/include
-                                                 data_override/include)
+                                                 data_override/include
```
[the remark this answers is not in the document]
Comment (J-Lentz) [reply CMakeLists.txt:326]: Fixed in d1924ee687dc6d4f1e226cb8ac0ac83bd8001169.

File: amip_interp/amip_interp.F90 [hunk 9]
```diff
@@ -146,18 +146,18 @@ module amip_interp_mod
...
+public amip_interp_init, amip_interp_init_r4, amip_interp_init_r8, get_amip_sst, &
+     & get_amip_ice, amip_interp_new, amip_interp_del, amip_interp_type_r4, &
+     & amip_interp_type_r8, assignment(=)
 
 !-----------------------------------------------------------------------
 !----------------- Public Data -----------------------------------
 integer :: i_sst = 1200
 integer :: j_sst = 600
-real, parameter:: big_number = 1.E30
+real(r8_kind), parameter:: big_number = 1.E30
```
[the remark this answers is not in the document]
Comment (J-Lentz) [reply amip_interp/amip_interp.F90]: Fixed in fb78dae60417e134dcdc2c8f2ac3181ab5fed580.

File: amip_interp/amip_interp.F90 (hunk 6 above)
Comment (J-Lentz) [reply amip_interp/amip_interp.F90]: `amip_interp_type` has been brought back in fb78dae60417e134dcdc2c8f2ac3181ab5fed580.

File: amip_interp/include/amip_interp.inc (hunk 7 above)
Comment (J-Lentz) [reply amip_interp/include/amip_interp.inc:146]: I think it's not possible to allow both `r4_kind` and `r8_kind` access in a way that's backward-compatible. We could make `sst_anom` and `sst_ncep` both `class(*), allocatable`, but then any user code that accesses them would have to be wrapped in a `select type` block.

File: amip_interp/amip_interp.F90 [hunk 10]
```diff
@@ -265,31 +275,39 @@ module amip_interp_mod
...
 type amip_interp_type
    private
-   type (horiz_interp_type) :: Hintrp, Hintrp2 ! add by JHC
-   real, allocatable        ::    data1(:,:),  data2(:,:)
-   type (date_type)         ::    Date1,       Date2
-   logical                  :: use_climo, use_annual
-   logical                  :: I_am_initialized=.false.
-end type
+   type (horiz_interp_type)              :: Hintrp, Hintrp2 ! add by JHC
+   class(*), dimension(:,:), allocatable :: data1, data2
```
Comment (mlee03) [inline amip_interp/amip_interp.F90:297]: I see pointers are being used in the .inc file cause there are class(*)? ...
Since data1 and data2 are private, can it be 
``real(r4_kind), dimension(:,:) :: data1_r4, data2_r4``
      ``real(r8_kind), dimension(:,:) :: data1_r8, data2_r8``

Then in the .inc file, the these arrays can be accessed by Interp%DATA1_KIND_ where DAT1_KIND_ is defined as data1_r4 in the .fh file.  @rem1776  came up with the idea and we've been doing something like that for other modules.

**Review by mcallic2 — APPROVED** [review]

File: amip_interp/amip_interp.F90 [hunk 11]
```diff
@@ -146,16 +146,17 @@ module amip_interp_mod
...
 !-----------------------------------------------------------------------
 !----------------- Public Data -----------------------------------
 integer :: i_sst = 1200
 integer :: j_sst = 600
-real, parameter:: big_number = 1.E30
+real(r8_kind), parameter:: big_number = 1.E30_r8_kind
 logical :: forecast_mode = .false.
-real, allocatable, dimension(:,:) ::  sst_ncep, sst_anom
+real(r8_kind), allocatable, dimension(:,:) ::  sst_ncep,    sst_anom    !< r8 versions of sst_ncep and sst_anom
+real(r4_kind), allocatable, dimension(:,:) ::  sst_ncep_r4, sst_anom_r4 !< r4 versions of sst_ncep and sst_anom
```
Comment (mlee03) [inline amip_interp/amip_interp.F90:159]: I think it might be best to go with real(r8_kind) sst_ncep, sst_anom and cast it to r4_kind if necessary in the procedures.  Using sst_ncep_r4 and sst_anom_r4 would require way too many changes to the components that is out of scope with this project for now.

Comment (J-Lentz) [reply amip_interp/amip_interp.F90]: `sst_ncep_r4` and `sst_anom_r4` removed in cf84442e4163db85472c07529130402009c6915c.

File: amip_interp/include/amip_interp.inc [hunk 12]
```diff
@@ -519,8 +180,8 @@ else
 
     mobs_sst = 1440;  nobs_sst = 720
 
-    call set_sst_grid_edges_daily(mobs_sst, nobs_sst)
-    call horiz_interp_new ( Interp%Hintrp2, lon_bnd, lat_bnd, &
+    call SET_SST_GRID_EDGES_DAILY_ (mobs_sst, nobs_sst)
+    call horiz_interp_new ( Interp%Hintrp2, LON_BND_, LAT_BND_, &
```
Comment (mlee03) [inline amip_interp/include/amip_interp.inc:184]: Curious as to how this work.  Following how lon_bnd and lat_bnd are set in amip_interp_init , they're always going to be in r8_kind.  Wondering what happens if lon_model and lat_model are r4_kind...

Comment (J-Lentz) [reply amip_interp/include/amip_interp.inc:184]: In that case, the module should be initialized with `amip_interp_init_r4` rather than `amip_interp_init` (which is just an alias for `amip_interp_init_r8`), so that `LON_BND_` and `LAT_BND_` refer to the `r4_kind` versions.

Comment (mlee03) [reply amip_interp/include/amip_interp.inc:184]: I think we're trying to minimize the number of changes required on the components side.  It would be nice if the users can just call amip_interp_init instead of making changes to their code to call amip_interp_init_r4

Comment (J-Lentz) [reply amip_interp/include/amip_interp.inc:184]: It's a similar problem to `data_override`, where the initialization routine allocates real data but doesn't take any real arguments. We could take a similar approach here, e.g.,`amip_interp_init(mode=r4_kind)` or `amip_interp_init(mode=r8_kind)`, but I don't think we can avoid user code having to explicitly specify the kind unless we make `tempamip`, `lon_bnd`, and `lat_bnd` all `r8_kind`. Or perhaps make `amip_interp_init` a no-op and have `amip_interp_new` do the actual initialization.

Comment (mlee03) [reply amip_interp/include/amip_interp.inc:184]: Would it bad to make those arrays r8_kind?  It seems like that would make things simpler..

Comment (J-Lentz) [reply amip_interp/include/amip_interp.inc:184]: We could, although I'm not sure what the performance implications (if any) would be.

**Review by mcallic2 — APPROVED** [review]

File: libFMS.F90 [hunk 13]
```diff
@@ -73,18 +73,13 @@ module fms
...
   use amip_interp_mod, only: fms_amip_interp_init         => amip_interp_init, &
-                             fms_amip_interp_get_amip_sst             => get_amip_sst, &
-                             fms_amip_interp_get_amip_ice             => get_amip_ice, &
+                             fms_amip_interp_get_amip_sst => get_amip_sst, &
+                             fms_amip_interp_get_amip_ice => get_amip_ice, &
                              fms_amip_interp_new          => amip_interp_new, &
                              fms_amip_interp_del          => amip_interp_del, &
-                             FmsAmipInterp_type => amip_interp_type, &
-                             assignment(=), &
-                             fms_amip_interp_i_sst        => i_sst, &
```
[the remark this answers is not in the document]
Comment (J-Lentz) [reply libFMS.F90]: They were removed because I made them private in the first iteration of this PR, and then I neglected to add them back in. `libFMS.F90` has been restored to its original state in 58923c69613ad497d58d16391a2b1db8fd43835e.

**Review by mlee03 — APPROVED** [review]

**Review by rem1776 — APPROVED** [review]
